src/main/java/org/apache/datasketches/theta/AnotBimpl.java

```diff
@@ -0,0 +1,266 @@
+   * Construct a new AnotB SetOperation on the java heap.  Called by PairwiseSetOperation.
+   *
+   * @param seedHash 16 bit hash of the chosen update seed.
+   */
+  AnotBimpl(final short seedHash) {
+    seedHash_ = seedHash;
+    reset();
+  }
+
+  @Override
+  public void setA(final Sketch skA) {
+    if (skA == null) {
+      reset();
+      throw new SketchesArgumentException("The input argument must not be null");
+    }
+    if (skA.isEmpty()) {
+      reset();
+      return;
+    }
+    //skA is not empty
+    checkSeedHashes(seedHash_, skA.getSeedHash());
+    empty_ = false;
+    thetaLong_ = skA.getThetaLong();
+    final CompactSketch cskA = (skA instanceof CompactSketch)
+        ? (CompactSketch) skA
+        : ((UpdateSketch) skA).compact();
+    hashArr_ = skA.isDirect() ? cskA.getCache() : cskA.getCache().clone();
+    curCount_ = cskA.getRetainedEntries(true);
+  }
+
+  @Override
+  public void notB(final Sketch skB) {
+    if (empty_ || (skB == null) || skB.isEmpty()) { return; }
+    //skB is not empty
+    checkSeedHashes(seedHash_, skB.getSeedHash());
+    final long thetaLongB = skB.getThetaLong();
+    thetaLong_ = Math.min(thetaLong_, thetaLongB);
+
+    //Build hashtable and removes hashes of skB >= theta
+    final int countB = skB.getRetainedEntries(true);
+    CompactSketch cskB = null;
+    UpdateSketch uskB = null;
+    final long[] hashTableB;
+    if (skB instanceof CompactSketch) {
+      cskB = (CompactSketch) skB;
+      hashTableB = convertToHashTable(cskB.getCache(), countB, thetaLong_, REBUILD_THRESHOLD);
+    } else {
+      uskB = (UpdateSketch) skB;
+      hashTableB = (thetaLong_ < thetaLongB)
+          ? convertToHashTable(uskB.getCache(), countB, thetaLong_, REBUILD_THRESHOLD)
+          : uskB.getCache();
+      cskB = uskB.compact();
+    }
+
+    //build temporary arrays of skA
+    final long[] tmpHashArrA = new long[curCount_];
+
+    //search for non matches and build temp arrays
+    final int lgHTBLen = simpleIntLog2(hashTableB.length);
+    int nonMatches = 0;
+    for (int i = 0; i < curCount_; i++) {
+      final long hash = hashArr_[i];
+      if ((hash != 0) && (hash < thetaLong_)) { //skips hashes of A >= theta
+        final int index = hashSearch(hashTableB, lgHTBLen, hash);
+        if (index == -1) {
+          tmpHashArrA[nonMatches] = hash;
+          nonMatches++;
+        }
+      }
+    }
+    hashArr_ = Arrays.copyOfRange(tmpHashArrA, 0, nonMatches);
+    curCount_ = nonMatches;
+    empty_ = (nonMatches == 0) && (thetaLong_ == Long.MAX_VALUE);
+  }
+
+  @Override
+  public CompactSketch getResult(final boolean reset) {
+    return getResult(true, null, reset);
+  }
+
+  @Override
+  public CompactSketch getResult(final boolean dstOrdered, final WritableMemory dstMem,
+      final boolean reset) {
+    final CompactSketch result =
+        CompactOperations.componentsToCompact(
+            thetaLong_, curCount_, seedHash_, empty_, true, false, dstOrdered, dstMem, hashArr_);
```

Comment:
it seems that hashArr_ is not a copy, which it should be unless reset is requested

Reply:
Good catch.  Fixed.